homeassistant/components/stream/__init__.py

```diff
@@ -0,0 +1,161 @@
+
+    if not stream_source:
+        raise HomeAssistantError("Invalid stream source.")
+
+    if options is None:
+        options = {}
+
+    keepalive = hass.data[DOMAIN][CONF_KEEPALIVE]
+    try:
+        streams = hass.data[DOMAIN][ATTR_STREAMS]
+        stream = streams.get(stream_source)
+        if not stream:
+            stream = Stream(hass, stream_source,
+                            options=options, keepalive=keepalive)
+            streams[stream_source] = stream
+        if not preload and not stream.access_token:
+            stream.access_token = generate_secret()
+            stream.start()
+        return hass.data[DOMAIN][ATTR_ENDPOINTS][fmt].format(
+            hass.config.api.base_url, stream.access_token)
+    except Exception:
+        raise HomeAssistantError('Unable to get stream')
+
+
+@bind_hass
+def get_stream(hass, stream_source):
+    """Return available stream."""
+    streams = hass.data[DOMAIN][ATTR_STREAMS]
+    return streams.get(stream_source)
+
+
+async def async_setup(hass, config):
+    """Set up stream."""
+    conf = config[DOMAIN] if config.get(DOMAIN, {}) else {}
+
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][CONF_KEEPALIVE] = conf.get(CONF_KEEPALIVE)
+    hass.data[DOMAIN][ATTR_ENDPOINTS] = {}
+    hass.data[DOMAIN][ATTR_STREAMS] = {}
+
+    # Setup HLS
+    hls_endpoint = await async_setup_hls(hass)
+    hass.data[DOMAIN][ATTR_ENDPOINTS]['hls'] = hls_endpoint
+
+    @callback
+    def shutdown(event):
+        """Stop all stream workers."""
+        for stream in hass.data[DOMAIN][ATTR_STREAMS].values():
+            stream.keepalive = False
+            stream.stop()
+        _LOGGER.info("Stopped stream workers.")
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, shutdown)
+
+    return True
+
+
+class Stream:
+    """Represents a single stream."""
+
+    # pylint: disable=dangerous-default-value
+    def __init__(self, hass, source, options={}, keepalive=False):
+        """Initialize a stream."""
+        self.hass = hass
+        self.source = source
+        self.options = options
+        self.keepalive = keepalive
+        self.access_token = None
+        self._container = None
+        self._thread = None
+        self._thread_quit = None
+        self._outputs = {}
+
+    @property
+    def container(self):
+        """Return container."""
+        return self._container
+
+    @property
+    def outputs(self):
+        """Return stream outputs."""
+        return self._outputs
+
+    def add_provider(self, provider):
+        """Add provider output stream."""
+        if not self._outputs.get(provider.format):
+            self._outputs[provider.format] = provider
+        return self._outputs[provider.format]
+
+    def start(self):
+        """Start a stream."""
+        import av
+        if self._thread is None or not self._thread.isAlive():
+            self._container = av.open(self.source, options=self.options)
```

Comment:
This looks like will do I/O, we should move it to `stream_worker` thread, otherwise, it may dead-lock our event-loop. Just happened when I am debugging.

Reply:
You are correct.  I will move it.